### PR TITLE
Cody: Add tests for Inline Assist

### PR DESCRIPTION
**Pull request by abeatrix**
Close https://github.com/sourcegraph/sourcegraph/issues/51940

This PR adds a test for the lineTracker function used by `Inline Assist`. It is a major component that tracks changes made in the editor where Cody is working on to make sure the correct lines of code are getting highlighted, decorated, and replaced.

#### ~SOS~

This is the first time I work on adding a new test for function that has dependencies, and I can't seem to get it to work when importing the `vscode` package and the lineTracker function. This is why I duplicated the origin function in the test file instead of using import. Any advice on how to resolve this is greatly appreciated 🙇 

Update: Resolved

## Test plan

The newly added tests have all passed.

Here is another PR for adding e2e test for Inline Assist: https://github.com/sourcegraph/sourcegraph/pull/52337

### CODE REVIEW

File: client/cody/src/services/InlineController.test.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    } else if (change.end.line - change.start.line > 0) {
+        addedLines -= change.end.line - change.start.line
+    }
+    const newStartLine = change.start.line > cur.start.line ? cur.start.line : cur.start.line + addedLines
+    const newRange = new Range(newStartLine, 0, cur.end.line + addedLines, 0)
+    return newRange
+}
+
+describe('lineTracker returns a new selection range based on changes made to the current doc', () => {
+    it('Return null if change occurs after the current selected range', () => {
```
Comment (dominiccooney) [inline client/cody/src/services/InlineController.test.ts:19]: This is a nit but the conjugation of "it/return" is not right, it would be "it/returns", ditto below.

File: client/cody/src/services/InlineController.test.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
+import { Range } from './VSCodeMocks'
+
+function lineTrackerTest(change: Range, cur: Range, changeText: string): Range | null {
```
Comment (dominiccooney) [inline client/cody/src/services/InlineController.test.ts:3]: Don't duplicate the function because that creates toil to keep them in sync and we are not really testing the code we are running for realz.

vscode imports fails because the unit tests run in a test harness, not in a full on vscode environment.

You have two choices:

Write an integration test (see test:integration), so you are running in vscode. That's the right answer if you're depending on a lot of machinery in vscode, not really the case here, you just want that range type.

So choice two, keep this unit test, but adjust the type of the function to work with VScode's ranges or "test" ranges. This is what you want to do here. Luckily TypeScript makes this easy because its types are described by their structure, not their names like they are in most languages.

Look at what I'm doing here: https://github.com/sourcegraph/sourcegraph/commit/a2ba1ae63f4db086929118e0118f96602986a40c#diff-39ceb083daa613399368b5642a1ecdf8e0fee3c7b94dd1edd30b11f9466dc198 defining a type which is a similar shape to the vscode position and range (the parts we care about) and then make the function accept any type of range: https://github.com/sourcegraph/sourcegraph/commit/a2ba1ae63f4db086929118e0118f96602986a40c#diff-af9b453968c3c3eb5ee7340cea67331775a8d458c2719b36d1b2f257e5ce25a1R32

There are simpler ways to do it depending on exactly what's going on. Because I was building a new data structure which contained one of the ranges I had a bunch of that <Self, ...> template stuff, but you should just be able to make a new type describing the fields of the range that you need.

Then change the function type so it can accept any kind of range.

LMK if you need more pointers.

File: client/cody/src/services/InlineController.ts [hunk 3]
```diff
@@ -73,9 +73,8 @@ export class InlineController {
             ) {
                 return
             }
-            const newRange = lineTracker(e, this.selectionRange)
-            if (newRange) {
-                this.selectionRange = newRange
+            for (const change of e.contentChanges) {
+                this.selectionRange = lineTracker(change.range, this.selectionRange, change.text) || this.selectionRange
```
Comment (dominiccooney) [inline client/cody/src/services/InlineController.ts:77]: Why not simply return the range unchanged and simplify this call site. There will probably be more callers over time... they will probably all do this.

File: client/cody/src/services/CodeLensProvider.ts [hunk 4]
```diff
@@ -28,18 +28,11 @@ export class CodeLensProvider implements vscode.CodeLensProvider {
                     this.remove()
                     return
                 }
-                let addedLines = 0
-                if (change.text.includes('\n')) {
-                    addedLines = change.text.split('\n').length - 1
-                } else if (change.range.end.line - change.range.start.line > 0) {
-                    addedLines -= change.range.end.line - change.range.start.line
+                const newRange = lineTracker(change.range, this.ranges, change.text)
```
Comment (dominiccooney) [inline client/cody/src/services/CodeLensProvider.ts:31]: lineTracker sounds like an object for tracking... lines... give this a descriptive verb phrase.

What about updateRangeForChangedLines or something like that?

File: client/cody/src/services/InlineController.ts [hunk 5]
```diff
@@ -332,26 +331,21 @@ export class Comment implements vscode.Comment {
...
-            addedLines = change.text.split('\n').length - 1
-        } else if (change.range.end.line - change.range.start.line > 0) {
-            addedLines -= change.range.end.line - change.range.start.line
-        }
-        const newRange = new vscode.Range(
-            new vscode.Position(cur.start.line + addedLines, 0),
-            new vscode.Position(cur.end.line + addedLines, 0)
-        )
-        return newRange
+export function lineTracker(change: vscode.Range, cur: vscode.Range, changeText: string): vscode.Range | null {
```
Comment (dominiccooney) [inline client/cody/src/services/InlineController.ts:336]: This API would be better if you put related things together. The changed text is more related to the changed range, so put them together. I could definitely image reading a callsite that says lineTracker(trackedRange, changedRange, changeText) and think that is fine, but it is wrong! Make the right thing to do the easiest and most natural thing to do.

I would even go as far as passing the text edit object which contains the range and the text. That way there's even less opportunity to mess up.

File: client/cody/src/services/InlineController.ts [hunk 6]
```diff
@@ -332,26 +331,21 @@ export class Comment implements vscode.Comment {
...
+    let addedLines = 0
+    if (changeText.includes('\n')) {
+        addedLines = changeText.split('\n').length - 1
+    } else if (change.end.line - change.start.line > 0) {
+        addedLines -= change.end.line - change.start.line
     }
-    return null
+    const newStartLine = change.start.line > cur.start.line ? cur.start.line : cur.start.line + addedLines
+    const newRange = new vscode.Range(newStartLine, 0, cur.end.line + addedLines, 0)
+    return newRange
```
Comment (dominiccooney) [inline client/cody/src/services/InlineController.ts:348]: It is neat that you don't need to validate the ranges because you're always using the 0th character of the line as the position... and all lines have at least zero characters.

File: client/cody/src/services/VSCodeMocks.ts [hunk 7]
```diff
@@ -0,0 +1,127 @@
+/* ------------------------------------------------------------------------------------------------
+ * A collection of mock classes for testings that require the vscode package.
+ *-----------------------------------------------------------------------------------------------*/
+
+// Mock vscode.Position
+export class Position {
```
Comment (dominiccooney) [inline client/cody/src/services/VSCodeMocks.ts:6]: I think we could share some code and interfaces with that WIP branch I pointed to earlier.

**Review by dominiccooney — COMMENTED** [review]
Good start, comments inline. I would work on de-duplicating the function definition first.

**Comment by abeatrix** [comment]
Updated the tests based on pointers from @dominiccooney and @valerybugakov so the test was not duplicated anymore, thank you 🙇

File: client/cody/src/services/InlineAssist.test.ts [hunk 8]
```diff
@@ -0,0 +1,83 @@
+import * as vscode from 'vscode'
+
+import { updateRangeOnDocChange } from './InlineAssist'
+
+jest.unmock('./InlineAssist')
```
Comment (valerybugakov) [inline client/cody/src/services/InlineAssist.test.ts:5]: Why is this required?

File: client/cody/src/services/InlineAssist.test.ts [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+jest.mock('vscode', () => {
```
Comment (valerybugakov) [inline client/cody/src/services/InlineAssist.test.ts:7]: Great! In a follow-up PR, we can consolidate this into a shared script so that other tests can utilize the same VSCode mocks. I anticipate the mocks will be uniform across all tests since we're required to mock all properties of vscode.

**Review by valerybugakov — APPROVED** [review]
💪

**Review by valerybugakov — APPROVED** [review]
💪

File: client/cody/src/services/InlineAssist.test.ts (hunk 8 above)
Comment (abeatrix) [reply client/cody/src/services/InlineAssist.test.ts]: Woops it was a leftover. Removed!
